# ANOVA Example

Suppose a bank has several branches, and each branch offers personal loans with varying interest rates. We want to understand if the mean interest rates are significantly different across these branches.

Groups for ANOVA: We will compare the mean interest rate offered by different branches. Let's assume the bank has three branches: Branch A, Branch B, and Branch C.

In [1]:
import pandas as pd
import scipy.stats as stats

In [3]:
df = pd.read_csv('banking_interest_rates.csv')
df

Branch  Interest_Rate
0   Branch A       6.205621
1   Branch A       5.660063
2   Branch A       5.891495
3   Branch A       6.396357
4   Branch A       6.247023
..       ...            ...
85  Branch C       6.558356
86  Branch C       6.271512
87  Branch C       5.728030
88  Branch C       5.371699
89  Branch C       6.221781

[90 rows x 2 columns]

In [4]:
A_interest = df[df['Branch'] == 'Branch A']['Interest_Rate']
B_interest = df[df['Branch'] == 'Branch B']['Interest_Rate']
C_interest = df[df['Branch'] == 'Branch C']['Interest_Rate']

In [5]:
F, p = stats.f_oneway(A_interest,B_interest, C_interest)
F, p

(2.097882588257023, 0.12888111157589846)

Interpret Results:

- If the p-value (p) is less than a significance level (commonly 0.05), it suggests that there are significant differences in the mean interest rates across the branches.
- If p is greater than this threshold, it suggests no significant difference.

So in this case, we can conclude that there is no significant difference between the interst rates of different branches

# Using other package to run ANOVA

In [7]:
df

Branch  Interest_Rate
0   Branch A       6.205621
1   Branch A       5.660063
2   Branch A       5.891495
3   Branch A       6.396357
4   Branch A       6.247023
..       ...            ...
85  Branch C       6.558356
86  Branch C       6.271512
87  Branch C       5.728030
88  Branch C       5.371699
89  Branch C       6.221781

[90 rows x 2 columns]

In [6]:
from scipy.stats import f_oneway
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Fitting the ANOVA model
model = ols('Interest_Rate ~ C(Branch)', data=df).fit()

# Performing ANOVA and getting the ANOVA table
anova_table = sm.stats.anova_lm(model, typ=2)

anova_table

sum_sq    df         F    PR(>F)
C(Branch)   0.666369   2.0  2.097883  0.128881
Residual   13.817291  87.0       NaN       NaN

#### Interpretation:

- **Sum of Squares (Sum_sq):** This column represents the total variation attributed to each source (between branches and within branches).
- **Degrees of Freedom (df):** Reflects the number of levels in the factor (for 'C(Branch)', it's 2, representing the three branches) and the residual degrees of freedom.
- **F value:** This is the calculated statistic for the ANOVA test. A higher F value usually indicates a more significant difference between the group means.
- **p-value (PR(>F)):** The p-value indicates the probability of observing the data if the null hypothesis is true. In this case, the p-value is approximately 0.129.

Since the p-value is greater than the common alpha level of 0.05, we fail to reject the null hypothesis. This suggests that there is no statistically significant difference in the mean interest rates offered by the three branches.

#### ols(Interest_Rate ~ C(Branch), data=df).fit

The formula `'Interest_Rate ~ C(Branch)'` used in the context of ANOVA with Python's `statsmodels` library is a way to specify the model being tested. Let's break down its components:

1. **`Interest_Rate`:** This is the dependent variable in the model. It represents the variable whose variation we are trying to explain. In your dataset, 'Interest_Rate' refers to the interest rates offered by different bank branches.

2. **`~`:** This symbol is a separator in the formula and is read as "modeled by". It separates the left-hand side (the dependent variable) from the right-hand side (the independent variable or factors).

3. **`C(Branch)`:** 
   - `Branch` is the independent variable or factor in the model. In this case, it represents the different branches of the bank.
   - The `C()` function is used to indicate that 'Branch' is a categorical variable


# Inference for Variance

## Inference for Variance in a Single Population

A bank wants to assess the consistency of loan approval times across its single branch. The goal is to determine if the variance in approval times is within an acceptable range.

Data Assumptions
- We have a sample of loan approval times (in days) from a single branch.
- The sample size is, let's say, 30.

We will use the Chi-Square test for a single variance, which tests whether the population variance equals a specified value based on a sample variance.

In [1]:
import numpy as np
from scipy.stats import chi2

# Sample data for loan approval times (asuming we have the data)
approval_times = np.array([3, 4, 2, 5, 4, 3, 5, 4, 3, 2, 4, 3, 5, 4, 3, 3, 4, 3, 5, 4, 3, 2, 4, 3, 5, 4, 3, 4, 3, 5])

In [2]:
# Sample variance
sample_variance = np.var(approval_times, ddof=1)

# Assuming a hypothesized population variance
hypothesized_variance = 4

Hypotheses
- Null Hypothesis (H0): The variance of loan approval times is equal to the hypothesized variance (σ² = 4 days²).
- Alternative Hypothesis (H1): The variance of loan approval times is not equal to the hypothesized variance (σ² ≠ 4 days²).

In [3]:
# Chi-square statistic
chi_square_statistic = (len(approval_times) - 1) * sample_variance / hypothesized_variance

In [6]:
chi_square_statistic 

6.241666666666666

In [7]:
# Degrees of freedom
df = len(approval_times) - 1
df

29

In [8]:
# P-value
p_value = chi2.sf(chi_square_statistic, df)

In [9]:
chi_square_statistic, p_value

(6.241666666666666, 0.9999975856768911)

**So what is your conclusion?**

## Inference for Variance in Two Populations

A bank operates two branches and wants to compare the variability in loan approval times between these two branches.

Data Assumptions:
- We have two independent samples of loan approval times from two branches.
- The sample sizes are not necessarily the same.

We will use the F-test for equality of two variances, which compares the variances of two independent samples.

In [10]:
# Sample data for two branches (Let's assume we have such data)
approval_times_branch1 = np.array([3, 4, 5, 2, 4, 5, 3, 4, 5, 3])
approval_times_branch2 = np.array([4, 4, 5, 5, 4, 5, 4, 3, 4, 5, 4, 5])

Hypotheses
- Null Hypothesis (H0): The variances in loan approval times are equal across the two branches (σ₁² = σ₂²).
- Alternative Hypothesis (H1): The variances in loan approval times are not equal across the two branches (σ₁² ≠ σ₂²).

In [11]:
from scipy.stats import f

# Sample variances
variance_branch1 = np.var(approval_times_branch1, ddof=1)
variance_branch2 = np.var(approval_times_branch2, ddof=1)

# F-statistic
F_statistic = variance_branch1 / variance_branch2

# Degrees of freedom for both samples
df1 = len(approval_times_branch1) - 1
df2 = len(approval_times_branch2) - 1

# P-value
p_value = 2 * min(f.cdf(F_statistic, df1, df2), 1 - f.cdf(F_statistic, df1, df2))

F_statistic, p_value

(2.514285714285715, 0.15173516176108715)

The reason why we should use `min` because it is the area of rejection area. This the way that we use to calculate p-value for non-symmetric distribution-based test

## Inference for Variance in >2 Populations

It is complicated!

- Bartlett’s test: Compare the variances of k samples, where k can be more than two samples. The data must be normally distributed. The Levene test is an alternative to the Bartlett test that is less sensitive to departures from normality.
You can follow https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bartlett.html

- Levene’s test: Compare the variances of k samples, where k can be more than two samples. It’s an alternative to the Bartlett’s test that is less sensitive to departures from normality.
You can follow https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.levene.html


- Fligner-Killeen test: a non-parametric test which is very robust against departures from normality.
You can follow https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fligner.html